jetty-server/src/main/java/org/eclipse/jetty/server/session/Session.java

```diff
@@ -127,31 +126,45 @@ public void onTimeoutExpired()
                 {
                     if (LOG.isDebugEnabled())
                         LOG.debug("Timer expired for session {}", getId());
-                    getSessionHandler().sessionInactivityTimerExpired(Session.this);
+                    long now = System.currentTimeMillis();
+                    //handle what to do with the session after the timer expired
+                    getSessionHandler().sessionInactivityTimerExpired(Session.this, now);
+                    try (Lock lock = Session.this.lock())
+                    {
+                        //grab the lock and check what happened to the session: if it didn't get evicted and
+                        //it hasn't expired, we need to reset the timer
+                        if (Session.this.isResident() && Session.this.getRequests() <= 0 && Session.this.isValid() && !Session.this.isExpiredAt(now))
+                        {
+                            //session wasn't expired or evicted, we need to reset the timer
+                            SessionInactivityTimer.this.schedule(Session.this.calculateInactivityTimeout(now));
+                        }
+                    }
                 }
 
             };
         }
 
-
         /**
-         * @param ms the timeout to set; -1 means that the timer will not be
-         *            scheduled
+         * For backward api compatibility only. 
+         * @see #schedule(long)
          */
-        public void setTimeout(long ms)
+        @Deprecated
+        public void schedule ()
         {
-            _msec = ms;
-            if (LOG.isDebugEnabled())
-                LOG.debug("Session {} timer={}ms", getId(), ms);
+            schedule(calculateInactivityTimeout(System.currentTimeMillis()));
         }
 
-        public void schedule()
+        /**
+         * @param ms the timeout to set; -1 means that the timer will not be
+         *            scheduled
+         */
+        public void schedule (long time)
         {
-            if (_msec > 0)
+            if (time >= 0)
```

Comment:
You don't want to schedule is `time==0`.

Reply:
I disagree. This is protection in case the scheduler is ever called with 0 time remaining: in this case, we want the usual codepath followed to handle expirations, and that is for the timer to call onTimeoutExpired.